lib/segment/src/index/field_index/full_text_index/inverted_index.rs

```diff
@@ -253,4 +275,215 @@ impl InvertedIndex {
+        self.get_doc(point_id).map(|x| x.len()).unwrap_or(0)
+    }
+
+    pub fn values_is_empty(&self, point_id: PointOffsetType) -> bool {
+        self.get_doc(point_id).map(|x| x.is_empty()).unwrap_or(true)
+    }
+
+    pub fn check_match(&self, parsed_query: &ParsedQuery, point_id: PointOffsetType) -> bool {
+        if let Some(doc) = self.get_doc(point_id) {
+            parsed_query.check_match(doc)
+        } else {
+            false
+        }
+    }
+
+    fn get_doc(&self, idx: PointOffsetType) -> Option<&Document> {
+        self.point_to_docs.get(idx as usize)?.as_ref()
+    }
+}
+
+#[derive(Default)]
+pub struct ImmutableInvertedIndex {
+    postings: Vec<Option<PostingList>>,
+    vocab: HashMap<String, TokenId>,
+    point_documents_tokens: Vec<Option<usize>>,
+    points_count: usize,
+}
+
+impl ImmutableInvertedIndex {
+    pub fn remove_document(&mut self, idx: PointOffsetType) -> bool {
+        if self.values_is_empty(idx) {
+            return false; // Already removed or never actually existed
+        }
+        self.point_documents_tokens[idx as usize] = None;
+        self.points_count -= 1;
+        true
+    }
+
+    pub fn values_is_empty(&self, point_id: PointOffsetType) -> bool {
+        if self.point_documents_tokens.len() <= point_id as usize {
+            return true;
+        }
+        self.point_documents_tokens[point_id as usize].is_none()
+    }
+
+    pub fn values_count(&self, point_id: PointOffsetType) -> usize {
+        if self.point_documents_tokens.len() <= point_id as usize {
+            return 0;
+        }
+        self.point_documents_tokens[point_id as usize].unwrap_or(0)
+    }
+
+    pub fn check_match(&self, parsed_query: &ParsedQuery, point_id: PointOffsetType) -> bool {
+        if parsed_query.tokens.contains(&None) {
+            return false;
+        }
+        // check presence of the document
+        if self.values_is_empty(point_id) {
+            return false;
+        }
+        // Check that all tokens are in document
+        parsed_query
+            .tokens
+            .iter()
+            // unwrap crash safety: all tokens exist in the vocabulary if it passes the above check
+            .all(|query_token| {
+                if let Some(posting_list) = &self.postings[query_token.unwrap() as usize] {
+                    posting_list.contains(&point_id)
+                } else {
+                    false
+                }
+            })
+    }
+}
+
+impl From<MutableInvertedIndex> for ImmutableInvertedIndex {
+    fn from(index: MutableInvertedIndex) -> Self {
+        ImmutableInvertedIndex {
+            postings: index.postings,
+            vocab: index.vocab,
+            point_documents_tokens: index
+                .point_to_docs
+                .iter()
+                .map(|doc| doc.as_ref().map(|doc| doc.len()))
+                .collect(),
+            points_count: index.points_count,
+        }
+    }
```

Comment:
Would it make sense here to to call [`shrink_to_fit`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.shrink_to_fit) on some of the types here, since we won't mutate them anyway?

Thinking about [this](https://github.com/qdrant/qdrant/issues/804#issuecomment-1931534082) here.

Reply:
It makes sense, fixed